Add forward AD formulas for some losses

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71026

Tracker: https://github.com/pytorch/pytorch/issues/71117

~...and fmod~

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- ~multi-margin-loss: new module tests~
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

soulitzer committed:
Add forward AD formulas for some losses

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33485661).

soulitzer committed:
Update on "Add forward AD formulas for some losses"


...and fmod

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- multi-margin-loss: new module tests
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for some losses"


...and fmod

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- multi-margin-loss: new module tests
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33485661).

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -631,10 +631,12 @@
 - name: fmod.Tensor(Tensor self, Tensor other) -> Tensor
   self: grad
   other: 'not_implemented("fmod: other")'
+  result: self_t.broadcast_to(result.sizes())
```
That is not correct? or you should have a check that ensures that `other` is always zeros.
Otherwise, we should implement both the forward and backward formulas related to other.

soulitzer replied:
Whoops, I think I need to study this operator more. For this PR, I will just revert this change since I realized @Lezcano already has PR out for this.

Lezcano replied:
For reference, I have one that does this one and it's cousin `torch.reminder`.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1791,8 +1800,6 @@
 
 - name: glu(Tensor self, int dim=-1) -> Tensor
   self: glu_backward(grad, self, dim)
-  # RuntimeError: output with shape [1] doesn't match the broadcast shape [2]
```
Should this be added back?

soulitzer replied:
Nope. I realized that this op isn't even element-wise... it actually splits the input in half and then applies binary-element-wise, so this error is expected if we attempt to do auto-element-wise

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -45,15 +45,13 @@ using at::areAnyTensorSubclassLike;
-      return unreduced.sum();
-    }
-    return unreduced;
+Tensor apply_loss_reduction(const Tensor& unreduced, int64_t reduction) {
```
Do we want to move it to VariabletTypeUtils then?

soulitzer replied:
I was thinking this belongs in the same category as maybe_multiply, handle_r_to_c etc?

albanD replied:
Yep.

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1394,14 +1392,26 @@ Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & in
+    if (!areAnyTensorSubclassLike({self, target})) {
+      grad_target = grad_output.mul(target.add(1).sub_(self).mul_(target.exp()));
+    } else {
+      grad_target = grad_output.mul(target.add(1).sub(self).mul(target.exp()));
```
The mul can remain inplace here no?

soulitzer replied:
Good point!

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1394,14 +1392,26 @@ Tensor kl_div_double_backward_grad_output(const Tensor & grad, const Tensor & in
+    if (!areAnyTensorSubclassLike({self, target}) && !grad_output._is_zerotensor()) {
+      grad_target = grad_output.mul(target.log().add_(1).sub_(self)).masked_fill_(target == 0, 0.);
+    } else {
+      grad_target = grad_output.mul(target.log().add(1).sub(self)).masked_fill(target == 0, 0.);
```
The masked_fill can remain inplace no?

soulitzer replied:
Oh whoops, it actually cannot if grad_output is a zero_tensor.

Lezcano commented:
Also, there's an open PR by @pmeier where he implements the OpInfo and a few other things for kl-div and the `l1_loss`.

pmeier commented:
See #69211.

soulitzer committed:
Update on "Add forward AD formulas for some losses"


Tracker: https://github.com/pytorch/pytorch/issues/71117

~...and fmod~

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- multi-margin-loss: new module tests
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer committed:
Update on "Add forward AD formulas for some losses"


Tracker: https://github.com/pytorch/pytorch/issues/71117

~...and fmod~

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- ~multi-margin-loss: new module tests~
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33485661).

albanD reviewed (approved):
Thanks!

soulitzer committed:
Update on "Add forward AD formulas for some losses"


Tracker: https://github.com/pytorch/pytorch/issues/71117

~...and fmod~

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- ~multi-margin-loss: new module tests~
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33485661).

soulitzer committed:
Update on "Add forward AD formulas for some losses"


Tracker: https://github.com/pytorch/pytorch/issues/71117

~...and fmod~

Testing:
- L1Loss: new module tests (linear in the real case only)
- SmoothL1Loss: new module tests
- MSELoss: tested - OpInfo + new module tests
- huberloss: tested - OpInfo + new module tests
- ~multi-margin-loss: new module tests~
- kl-div: OpInfo + new module tests
- fmod: OpInfo

Differential Revision: [D33485661](https://our.internmc.facebook.com/intern/diff/D33485661)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33485661).
